Fix

ILNCHV5 opened the pull request.

ILNCHV5 committed:
proptypes in progress

ILNCHV5 committed:
working, before proptypes

ILNCHV5 committed:
finished changes

ILNCHV5 committed:
fixed path

ILNCHV5 committed:
imports optimized

ivo96 commented on src/App.js:
```diff
@@ -1,69 +1,13 @@
+
+function App(){
+    return(
+        <Box>
```
You don't need Box here. Think in components not in html

ivo96 commented on src/api/usersApi.js:
```diff
@@ -0,0 +1,6 @@
+async function fetchUserData() {
```
You can use destructuring here for the json data like:
`const { json } = await ...` and return it.

ivo96 commented on src/api/usersApi.js:
```diff
@@ -0,0 +1,6 @@
+    return userData.json();
+  };
+
+export default fetchUserData;
```
We need named export not default, because you can have more than one function here right.
Use default for components only and named for functions which can be used throughout the project

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+import { React, useState } from "react";
```
We usualy import it like:
import React, { useState } from "react";

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+import { React, useState } from "react";
+import { Card, IconButton, Typography, CardContent, CardActions, Box } from "@mui/material";
```
When there are many imports we want to have them line by line, so it's more readable and maintainable

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+import { React, useState } from "react";
+import { Card, IconButton, Typography, CardContent, CardActions, Box } from "@mui/material";
+import { Favorite, Delete } from "@mui/icons-material";
+import styles from "../UserCard/cardStyles";
```
the file name can be just styles.css, because it's in the same folder

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+import styles from "../UserCard/cardStyles";
+import PropTypes from 'prop-types'
+
+export function UserCard({ id, name, email, phone, website, users, setUsers, favorites, setFavorites, isFav }) {
```
This is a lot of props. So what we want to do here is have an object for the card as a single prop and handler function if we need to do something with the data. That way it looks more simpler and easier on the eye

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+export function UserCard({ id, name, email, phone, website, users, setUsers, favorites, setFavorites, isFav }) {
+
+    const cardAvatar = `https://avatars.dicebear.com/v2/avataaars/{${name}}.svg?options[mood][]=happy`;
+    const [favStatus, setFavStatus] = useState(isFav);
```
Why do you need this? Just use the value of isFav prop in JSX directly

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+    const [favStatus, setFavStatus] = useState(isFav);
+
+    const addToFavs = () => {
+        if (favorites.indexOf({ id: id }) !== 1) {
```
What is the meaning of this if? Remove this and use the id of the user to iterate the favorites and change it's favorite property to its opposite. This should all be done in the setState. 
hint: use map

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+    };
+
+    const removeFromFavs = () => {
+        const filteredFavs = favorites.filter((item) => (item.id !== id));
```
Use the filter directly in setState - 
`setFavorites(favorites.filter...)`

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+        setFavStatus(false);
+    };
+
+    const handleFavs = (favStatus ? removeFromFavs : addToFavs);
```
`handleFavoritesButtonClick` is a better name
also just add the logic for adding and removing here you don't need the other 2 functions

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+
+    const handleFavs = (favStatus ? removeFromFavs : addToFavs);
+
+    const handleDelete = () => {
```
`handleDeleteButtonClick` is a better name

ivo96 replied:
Also, the `handleDeleteButtonClick` and `handleFavoriteButtonClick` should be a prop for this component. We want the component to just display the card.

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+                    <Box sx={styles.gradientBoxStyle} />
+                </Box>
+                <CardContent sx={styles.cardContentStyle}>
+                    <Typography component='div' sx={styles.cardTypographyStyle}>
```
Remove the 'div'. It should be something else..

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,77 @@
+        <Box>
+            <CardActions sx={styles.cardActionStyle}>
+                <IconButton size="large" sx={styles.FavIconStyle} onClick={handleFavs}>
+                    <Favorite color={favStatus ? 'action' : 'gray'} />
```
use the prop

ivo96 commented on src/components/UserCard/cardStyles.js:
```diff
@@ -1,3 +1,4 @@
+import { textAlign } from "@mui/system";
```
You don't need that remove it

ivo96 commented on src/customHooks/useLocalStorage.js:
```diff
@@ -0,0 +1,14 @@
+ 
+const retrievedFavs = JSON.parse(localStorage.getItem("favoriteCards"));
+
+export default function useLocalStorage(initialState) {
```
You don't need this initial state here

ivo96 commented on src/customHooks/useLocalStorage.js:
```diff
@@ -0,0 +1,14 @@
+const retrievedFavs = JSON.parse(localStorage.getItem("favoriteCards"));
+
+export default function useLocalStorage(initialState) {
+    const [favorites, setFavorites] = useState(retrievedFavs ? retrievedFavs : initialState);
```
this should only be `useState([])` and set it in the useEffect

ivo96 commented on src/customHooks/useLocalStorage.js:
```diff
@@ -0,0 +1,14 @@
+        localStorage.setItem("favoriteCards", favsToBeStored);
+      }, [favorites]);
+
+    return [favorites, setFavorites];
```
We don't want this state exported, we want functions to get all of the favorites and set one favorite (hint: use id).

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+import { React, useState, useEffect } from 'react';
```
import React, {useState ...}

ivo96 replied:
Rename the page with uppercase letter. If we have a component or page - it should be capitalized

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  return (
+    <ThemeProvider theme={theme}>
```
The Theme should have all of the pages, so this is not for here. Add the theme one level above in the App.

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+import useLocalStorage from '../../customHooks/useLocalStorage';
+
+function CardsPage() {
+
```
Remove unnecessary lines

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+import { UserCard } from '../../components/UserCard/UserCard';
+import theme from './theme';
+import './CardsPageStyles.css';
+import fetchUserData from '../../api/usersApi';
```
desctructuring here

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+  const [favorites, setFavorites] = useLocalStorage([]);
+
+  useEffect(() => {
+    fetchUserData().then(fetchedData => setUsers(fetchedData));
```
use async/await and try/catch instead of .then and .catch

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+    fetchUserData().then(fetchedData => setUsers(fetchedData));
+  }, []);
+
+  const handleInputChange = (event) => {
```
`handleSearchUser` is a better name

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+    setSearch(event.target.value);
+  }
+
+  const filterUsers = (users, search) => {
```
You don't need this function. Add the logic for the searching in the handleSearch handler.
 Also just manipulate one array of users which are displayed which in your case is the users state. IE you should have setUsers(...something here) in the handler

ILNCHV5 replied:
Not sure if its supposed to work the way i did it now, since i needs to refresh the page in order to reset the search, but i found no other way to do it with setUsers(). Would need a few more pointers if that's not right.

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+
+  const filterUsers = (users, search) => {
+    return users.filter((user) => {
+      if (users.length !== 0 && search === '') {
```
you don't need this here

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+      } else if (user.name.toLowerCase().includes(search.toLowerCase())) {
+        return user;
+      } else {
+        return null;
```
Why are we returning null, this is not right, remove it.

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+            website={user.website}
+            users={users} setUsers={setUsers}
+            favorites={favorites} setFavorites={setFavorites}
+            isFav={favorites.find(item => item.id === user.id) ? true : false} />))}
```
You don't need this.
just add isFavorite property to every object in the array when you fetch it.
Don't use find here.

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,59 @@
+
+  return (
+    <ThemeProvider theme={theme}>
+      <Box className="CardsPage">
```
Be consistent with the way you are describing classes. 
Either use styles (js) in all of the project or just import the css, but do only one of the above don't mix it. The key is **consistency**!! 
Also be consistent with the naming if you are using `camelCase` in all of the classes don't have the `CamelCase` also. We usualy have the first letter capitalised only for components. Again **consistency** !

ivo96 commented on src/pages/cardsPage/CardsPageStyles.css:
```diff
@@ -1,4 +1,4 @@
-.App {
+.CardsPage {
   display: flex;
   padding-top: 4rem;
```
Either use inline or css, don't use both.

ivo96 commented on src/customHooks/useLocalStorage.js:
```diff
@@ -0,0 +1,14 @@
+import { useState, useEffect } from 'react'
```
Folder should be just called hooks

ivo96 commented on src/pages/cardsPage/CardsPageStyles.css:
```diff
@@ -1,4 +1,4 @@
-.App {
+.CardsPage {
```
This is a container not a page

ivo96 commented:
finished changes is not a good name for commit. We have to know what is that commit changing and why(if it needs explanation)

ILNCHV5 committed:
worked on all comments, took the UserCard functions out of the card,
remade the search so it manipulates only one array,
and remade the favorites logic so they are assigned as a property
to all objects on setUser()

ILNCHV5 committed:
removed unused depencency

ILNCHV5 committed:
fixed repetitive code in cardsPage' useEffect

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,59 @@
+            </CardActions>
+            <Card sx={styles.cardStyle}>
+                <Box sx={styles.imagesContainerStyle}>
+                    <img style={styles.cardMediaStyle} src={cardAvatar} alt='cardAvatar'></img>
```
You don't need closing tag. Remove it

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,59 @@
+                <CardContent sx={styles.cardContentStyle}>
+                    <Typography component={Box} sx={styles.cardTypographyStyle}>
+                        <Box>Name</Box>
+                        <Box>{name}</Box> </Typography>
```
Use a code formatter, linter or prettifier to align your code. So it's more readable

ivo96 commented on src/components/UserCard/UserCard.jsx:
```diff
@@ -0,0 +1,59 @@
+}
+
+UserCard.propTypes = {
+    id: PropTypes.number,
```
I can see the props are user, handleDeleteButtonClick, handleFavoritesButtonClick. So I expect the proptypes to be again these three. Fix the proptypes. Also are these props required or not? read docs: https://reactjs.org/docs/typechecking-with-proptypes.html

ivo96 commented on src/hooks/useLocalStorage.js:
```diff
@@ -0,0 +1,18 @@
+    localStorage.setItem("favoriteCards", JSON.stringify(favorites));
+  }, [favorites]);
+
+  function toggleFavorite(id, isFavorite) {  
```
You don't need isFavorite(you should have just added it to each object prior), you can just use id, map the objects find the object which you are looking for and set it's favorite to the opposite.

ivo96 commented on src/hooks/useLocalStorage.js:
```diff
@@ -0,0 +1,18 @@
+  }, [favorites]);
+
+  function toggleFavorite(id, isFavorite) {  
+    setFavorites(favorites.find(item => item.id === id)
```
You are wasting the cpu and logic here, because you first find the object but then you map again all of the objects to find it again. It doesn't make sense to do that. You can just map all of the objects -> find the object you are looking for -> change whatever you need.

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,55 @@
+import React, { useState, useEffect } from 'react';
```
Try implementing this first without the localStorage hook. Don't save the state to the localStorage, because I think it confuses you.

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,55 @@
+    (async () => {
+      const userData = await fetchUserData();
+      const dataSource = users.length === 0? userData: users;
+      setUsers(dataSource.map(user => ({ ...user, isFavorite: checkIfFavorite(user.id) })));
```
Just use the fethedData and add the isFavorite to the users state. Implement everything without the localStorage and use only the users state to display the users. After that we'll discuss the localStorage thing.

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,55 @@
+  }, [favorites]);
+
+  const checkIfFavorite = (id) => {
+    const favorite = favorites.find(user => user.id === id);
```
No need for this here. Just use a map it's you best tool to iterate and do stuff to objects. You are doing twice the work for something simple.

ivo96 commented on src/pages/cardsPage/CardsPage.jsx:
```diff
@@ -0,0 +1,55 @@
+
+  const checkIfFavorite = (id) => {
+    const favorite = favorites.find(user => user.id === id);
+    return favorites.find(user => user.id === id) ? favorite.isFavorite : false;
```
Another way of solving this 
```suggestion
    return users
             .map(user => { 
                if (user.id === id) 
                  return user.isFav 
                })
              .some(user => user === true)
```

ILNCHV5 committed:
working without local storage of favorites

ILNCHV5 committed:
fixed useLocalStorage hook, so it works with just the id, removed excess processes.

ILNCHV5 commented:
The first commit is as you asked - without local storage functionality, in the second one i think i fixed the issues.
